ClavaLaraApi/src-lara-clava/clava/clava/pass/SingleReturnFunction.js

```diff
@@ -0,0 +1,41 @@
+laraImport("lara.pass.Pass");
+laraImport("weaver.Query");
+
+class SingleReturnFunction extends Pass {
+  constructor() {
+    super("SingleReturnFunction");
+  }
+
+  _apply_impl($jp) {
+    if ($jp.joinpointType !== "function" || !$jp.isImplementation) {
+      return;
+    }
+    const $body = $jp.body;
+    const $returnStmts = Array.from(Query.searchFrom($body, "returnStmt"));
+    if ($returnStmts.length <= 1) {
+      return;
+    }
+
+    $body.insertEnd("__returnLabel:");
+
+    const returnType = $jp.returnType;
+    const returnIsVoid =
+      !returnType.isBuiltin || returnType.builtinKind !== "Void";
+    if (returnIsVoid) {
+      $body.addLocal("__returnValue", returnType);
+      $body.insertEnd("return __returnValue;");
+    } else {
+      $body.insertEnd(";");
+    }
+
+    for (const $returnStmt of $returnStmts) {
+      if (returnIsVoid) {
+        $returnStmt.insertBefore(
+          `__return_value = ${$returnStmt.returnExpr.code};`
+        );
+      }
+      $returnStmt.insertBefore("goto __returnLabel;");
```

Comment:
This applies to any .insert... in the code, but is not a required change, just a suggestion for a future change. It is much more convenient to insert code as strings, however this requires a call to Clava.rebuild() if we want to transform the strings back to nodes (which implies reparsing all the code).

If a Pass can be implemented using only insertion of nodes, a call to Clava.rebuild() should not be needed. However, it is more cumbersome to directly create the nodes with clava.ClavaJoinPoints. Also, ClavaJoinPoints does not contains methods for all the nodes (e.g., goto, label), although they can be easily added.

Reply:
Will need to add the nodes to ClavaJoinPoints, but should be possible to avoid rebuilding the tree